lib/browser/desktop-capturer.js

```diff
@@ -10,37 +10,34 @@ const deepEqual = (a, b) => JSON.stringify(a) === JSON.stringify(b)
 // A queue for holding all requests from renderer process.
 let requestsQueue = []
 
-const electronSources = 'ELECTRON_BROWSER_DESKTOP_CAPTURER_GET_SOURCES'
-const capturerResult = (id) => `ELECTRON_RENDERER_DESKTOP_CAPTURER_RESULT_${id}`
-
-ipcMainInternal.on(electronSources, (event, captureWindow, captureScreen, thumbnailSize, fetchWindowIcons, id) => {
+ipcMainUtils.handle('ELECTRON_BROWSER_DESKTOP_CAPTURER_GET_SOURCES', (event, captureWindow, captureScreen, thumbnailSize, fetchWindowIcons) => {
   const customEvent = eventBinding.createWithSender(event.sender)
   event.sender.emit('desktop-capturer-get-sources', customEvent)
 
   if (customEvent.defaultPrevented) {
-    event._replyInternal(capturerResult(id), [])
-    return
+    return []
   }
 
-  const request = {
-    id,
-    options: {
-      captureWindow,
-      captureScreen,
-      thumbnailSize,
-      fetchWindowIcons
-    },
-    event
-  }
-  requestsQueue.push(request)
-  if (requestsQueue.length === 1) {
-    desktopCapturer.startHandling(captureWindow, captureScreen, thumbnailSize, fetchWindowIcons)
-  }
+  return new Promise((resolve, reject) => {
+    const request = {
+      options: {
+        captureWindow,
+        captureScreen,
+        thumbnailSize,
+        fetchWindowIcons
+      },
+      resolve
+    }
+    requestsQueue.push(request)
+    if (requestsQueue.length === 1) {
+      desktopCapturer.startHandling(captureWindow, captureScreen, thumbnailSize, fetchWindowIcons)
+    }
 
-  // If the WebContents is destroyed before receiving result, just remove the
-  // reference from requestsQueue to make the module not send the result to it.
-  event.sender.once('destroyed', () => {
-    request.event = null
+    // If the WebContents is destroyed before receiving result, just remove the
+    // reference from requestsQueue to make the module not send the result to it.
+    event.sender.once('destroyed', () => {
+      request.resolve = null
```

Comment:
Should `request` be removed from the `requestsQueue`?
Removing can be done easier if the `requestsQueue` were a `Set` instead of `Array`.

Reply:
let's leave it for now. queue kind of needs to be an array, not set